substrate-node/pallets/pallet-smart-contract/src/migration.rs

```diff
@@ -89,3 +184,233 @@ pub fn migrate_to_version_6<T: Config>() -> frame_support::weights::Weight {
         return 0;
     }
 }
+
+pub fn remove_deployment_contracts_from_contracts_to_bill<T: Config>() {
+    // we only bill capacity reservation contracts and name contracts
+    for index in 0..BillingFrequency::<T>::get() {
+        let mut contract_ids = ContractsToBillAt::<T>::get(index);
+        contract_ids.retain(|&id| {
+            let contract = Contracts::<T>::get(id).unwrap();
+            !matches!(contract.contract_type, types::ContractData::DeploymentContract(_))
+        });
+        ContractsToBillAt::<T>::insert(index, &contract_ids);
+    }
+}
+
+pub fn add_active_node_contracts<T: Config>() {
+    // active node contracts contains only capacity reservation contracts
+    ActiveNodeContracts::<T>::drain();
+    for (_contract_id, contract) in Contracts::<T>::iter() {
+        match contract.contract_type {
+            types::ContractData::CapacityReservationContract(crc) => {
+                let mut contracts = ActiveNodeContracts::<T>::get(crc.node_id);
```

Comment:
Why is this storage map drained before being read here?

Doc: "drain() - remove all elements from the map and iterate through them in no particular order. If you add elements to the map while doing this, you'll get undefined results."

I would either collect all the items, drop the storage map and start repopulating after.

see: https://github.com/threefoldtech/tfchain/blob/3e4caecf856b4f891e50bd274b69928b9de67fd4/substrate-node/pallets/pallet-smart-contract/src/migration.rs#L51

Reply:
We only need capacity reservation contracts in that storage map after these changes. So I thought I'd just remove all of them and go through the contracts to add only the capacity reservation contracts... Don't they mean if you add items while executing drain you'll get undefined results? Not if you do drain and then add back the items...

Reply:
How do you drop a storage map?

Reply:
how I read the code is:

1: first you drain the storage map (remove all items)
2: you try to read a key from the map (key won't exists because you deleted it)

Reply:
https://github.com/threefoldtech/tfchain/blob/016365c1ae964b72b5fe944a3aa8d3de98b7d2da/substrate-node/pallets/pallet-smart-contract/src/migration.rs#L54

Reply:
We never check the existence of keys in the storage map ActiveNodeContracts so I assume it returns an empty list in that case. Then we populate it back with new contract ids. So I think I should work but I'll use remove_storage_prefix.

Reply:
Hmm, I was wrong. I misinterpreted the code. You can drain here.